File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq
from collections import defaultdict
from dataclasses import dataclass
from collections import deque

@dataclass
class Job:
    id: int
    req_time : int
    duration : int

def solution(jobs):
    total_time = 0 # 총 반환시간
    done  = 0
    n = len(jobs)
    current_time = 0
    
    jobs_list = []
    for i, (req, dur) in enumerate(jobs):
        jobs_list.append(Job(i, req, dur))
    jobs_list.sort(key = lambda job : job.req_time)
    
    job_queue = deque(jobs_list)
    ready_queue = []
    
    while done < n:
        # 모든 작업 처리할 때까지
        while job_queue and job_queue[0].req_time <= current_time:
            job = job_queue.popleft()
            heapq.heappush(ready_queue, (job.duration, job.req_time, job.id, job))
        
        if ready_queue:
            # 대기큐에 있으면
            _, _, _, task = heapq.heappop(ready_queue)
            current_time += task.duration
            total_time += (current_time - task.req_time)
            done += 1
        else:
            current_time = job_queue[0].req_time
        
    
    return total_time // n

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

class Task{
    int id, reqTime, duration;
    
    public Task(int id, int reqTime, int duration){
        this.id = id;
        this.reqTime= reqTime;
        this.duration = duration;
    }
}
class Solution {
    public int solution(int[][] jobs) {
        // job을 요청사항 우선으로 정렬
        List<Task> jobList = new ArrayList<>();
        for(int i = 0; i < jobs.length; i++){
            jobList.add(new Task(i, jobs[i][0], jobs[i][1]));
        }
        
        jobList.sort(Comparator.comparingInt((Task t) -> t.reqTime));
        
        ArrayDeque<Task> jobQueue = new ArrayDeque<>(jobList);
        PriorityQueue<Task> readyQueue = new PriorityQueue<>(Comparator.comparingInt((Task t) -> t.duration)
                                                            .thenComparingInt(t -> t.reqTime)
                                                            .thenComparingInt(t -> t.id));
        
        int done = 0;
        int totalTime= 0;
        int current = 0;
        while(done < jobs.length){
            // current 전에 해당하는 요청사항 readyQueue에 넣기
            while(!jobQueue.isEmpty() && jobQueue.peekFirst().reqTime <= current){
                readyQueue.offer(jobQueue.removeFirst());
            }
            
            // readyQueue에 뭐 있으면 진행 없으면 요청사항 다음 것으로 current 이동
            if(!readyQueue.isEmpty()){
                Task task = readyQueue.poll();
                current += task.duration;
                totalTime += (current - task.reqTime);
                done++;
            }else{
                if(!jobQueue.isEmpty()) current = jobQueue.peekFirst().reqTime;
            }
        }
        
        return totalTime/jobs.length;
    }
}